### PR TITLE
Add ability to close playlists within grace period after creation

### LINKED ISSUE #30401: Ability to close playlists within time period after creation
It's quite easy to make a mistake and want to recreate a playlist. I've had several requests for manual deletion for cases like this.

### Discussed in https://github.com/ppy/osu/discussions/30393

<div type='discussions-op-text'>

<sup>Originally posted by **Hiviexd** October 23, 2024</sup>
would be a very appreciated UX improvement imo. 

in multiple past cases related to spotlights/loved (most recently in [osu!dev](<https://discord.com/channels/188630481301012481/384640068847009792/1298386123210625165>)), we had to request manual edits/deletions to playlists due to mistakes during the creation process (wrong mods, title typo, wrong deadline). so it would be great if users were allowed to close their playlists within 5 minutes after creation, that'd help with solving any potential mistakes while preserving the long-term integrity of playlist leaderboards.</div>

### PR DESCRIPTION
**Pull request by bdach**
Closes https://github.com/ppy/osu/issues/30401.

https://github.com/user-attachments/assets/33ad7c58-b2c0-4c35-81f3-4e6e54d29bb8

Grace period in question is configurable osu-web side via `MULTIPLAYER_ROOM_CLOSE_GRACE_PERIOD_MINUTES` envvar; default is 5 minutes (see https://github.com/ppy/osu-web/pull/11667). Same value is hardcoded in client.

Not sure the lounge context menu is the best place for this necessarily but that was kinda simplest...?

No tests because it felt like adding them would be too expensive for the benefit, but will add on request.

### CODE REVIEW

**Comment by peppy** [comment]
I think we might want to, in addition to the context menu, show a button next to "Start" on the playlist screen for discoverability. What do you think?

**Comment by bdach** [comment]
Yeah I considered doing that. Not opposed to it in principle, will give it a go.

**Comment by bdach** [comment]
osu_2024-11-22_09-59-16

Something like this I guess...?

**Comment by peppy** [comment]
Just thinking about this one, the number one case where someone is looking to close a playlist is where they made a mistake and wish to make an edit. Do you think it would be good to have an "Edit" button as well (which clones the playlist and places the user back at the create screen)... and maybe even rename "Close" to "Delete" since that may make more sense to an end user?

**Comment by peppy** [comment]
This flow feels a bit weird:

https://github.com/user-attachments/assets/81ec5da1-5715-41ef-b2ea-bb969788f58e

I'd hope for the "Start" button to be disabled at the point of closing, even if we need to shortcut this to happen without a response from the server implying as such.

File: osu.Game/Screens/OnlinePlay/Playlists/ClosePlaylistDialog.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+using osu.Game.Online.Rooms;
+using osu.Game.Overlays.Dialog;
+
+namespace osu.Game.Screens.OnlinePlay.Playlists
+{
+    public partial class ClosePlaylistDialog : DeletionDialog
+    {
+        public ClosePlaylistDialog(Room room, Action closeAction)
+        {
+            HeaderText = "Are you sure you want to close the following playlist:";
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Playlists/ClosePlaylistDialog.cs:14]: I hesitantly mentioned this in my other comment, but we could just expose this to the user as "deleting" and not require custom terminology for closing. I think to the user they expect to "delete" a playlist, not "close" it.. but I can swing both ways.

File: osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs [hunk 2]
```diff
@@ -2,32 +2,113 @@
...
+
+        private void hideCloseButton()
+        {
+            closeButton?.ResizeWidthTo(0, 100, Easing.OutQuint)
+                       .Then().FadeOut().Expire();
+        }
+
+        private void onRoomChanged(object? sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == nameof(Room.Status) || e.PropertyName == nameof(Room.Host) || e.PropertyName == nameof(Room.StartDate))
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs:86]: Kinda prefer this written as a `switch` when there's this many properties being matched:

```diff
diff --git a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs
index c8e5f0859d..9ccc8f3cab 100644
--- a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs
+++ b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomFooter.cs
@@ -83,8 +83,14 @@ private void hideCloseButton()
 
         private void onRoomChanged(object? sender, PropertyChangedEventArgs e)
         {
-            if (e.PropertyName == nameof(Room.Status) || e.PropertyName == nameof(Room.Host) || e.PropertyName == nameof(Room.StartDate))
-                updateState();
+            switch (e.PropertyName)
+            {
+                case nameof(Room.Status):
+                case nameof(Room.Host):
+                case nameof(Room.StartDate):
+                    updateState();
+                    break;
+            }
         }
 
         private void updateState()

```

**Comment by bdach** [comment]
> Just thinking about this one, the number one case where someone is looking to close a playlist is where they made a mistake and wish to make an edit. Do you think it would be good to have an "Edit" button as well (which clones the playlist and places the user back at the create screen)...

You can create a copy of a closed room which would kinda cover this usecase. Not sure I'd want to add anything specially intended for this feature other than that.

File: osu.Game/Screens/OnlinePlay/Playlists/ClosePlaylistDialog.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Playlists/ClosePlaylistDialog.cs:14]: Am neither here or there, fine with either personally.

**Comment by bdach** [comment]
> I'd hope for the "Start" button to be disabled at the point of closing, even if we need to shortcut this to happen without a response from the server implying as such.

Fixed in https://github.com/ppy/osu/pull/30793/commits/c93c549b054a7e4ec5935a05e9bb71be807a5182. Bit ad hoc but seems to work well enough.

**Review by peppy — APPROVED** [review]

**Pull request merged**